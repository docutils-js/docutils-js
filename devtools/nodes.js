#!/usr/bin/env node
const recast = require('recast');
const j = require('jscodeshift');
const fs = require('fs');
const path = require('path');
const PropTypes = require('prop-types');
const nodes = require('../lib/nodes');
const nodesSrc = path.join(__dirname, '../src/nodes.js');
// const nodesJs = fs.createReadStream(nodesSrc);
const source = fs.readFileSync(nodesSrc, { encoding: 'utf-8' });
const file = recast.parse(source);

const superClasses = [];
const leafClasses = [];
let classes = ['Node'];
const r = j(source);
const paths = r.find(j.ClassDeclaration, p => p.superClass).paths();
const nodesInfo = {};
let curNodeInfo;
const flatNodesInfo = {};
/* Each iteration we process the Nth level of the class hierarchy, starting from the root class Node */
while (classes.length) {
    const nextClasses = [];
    classes.forEach((superClass) => {
        curNodeInfo = nodesInfo[superClass];
        const subClasses = paths.filter(p => p.value.superClass.name === superClass).map(p => p.value.id.name);
        const myInfo = {};
        myInfo.leaf = subClasses.length === 0;
        const args = [];
        if (subClasses.length) {
            myInfo.subClasses = subClasses;
            superClasses.push(superClass);
        } else {
            leafClasses.push(superClass);
            const params = nodes[superClass].params;
            if (params) {
                if (Array.isArray(params)) {
                    params.forEach(param => {
                        if (param == nodes.Params.ElementContent) {
                            args.push('Element Content');
                        } else if (param == nodes.Params.RawSource) {
                            args.push('Raw Source');
                        }
                    });
                }
                ;
            }
            //PropTypes.checkPropTypes(nodes[superClass].propTypes, {});
            console.log(superClass);
        }
        try {
            const node = new nodes[superClass](...args);
            myInfo.classTypes = node.classTypes.map(c => c.name);
        } catch(err) {
            myInfo.error = err.message;
        }
//            if(!node.test()) {
//                throw new Error();;
//            }
            myInfo.name = superClass;
            flatNodesInfo[myInfo.name] = myInfo;

        nextClasses.splice(nextClasses.length, 0, ...subClasses);
        console.log(myInfo);
    });
    classes = nextClasses;
}
fs.writeFileSync('nodes.json', JSON.stringify(flatNodesInfo));

console.log(JSON.stringify(flatNodesInfo));
const transBody = j.classBody([]);
const file2 = recast.parse('import BaseWriter from \'../Writer\';\n'
                           + 'import { GenericNodeVisitor } from \'../nodes\';\n');
const classDecl = j.classDeclaration(j.identifier('Translator'), transBody, j.identifier('GenericNodeVisitor'));
file2.program.body.push(classDecl);
const f = file2;
// j.program([j.importDeclaration([j.importNamespaceSpecifier(j.identifier('BaseWriter'))],
//    j.literal('../Writer'), ]));
// console.log(Object.keys(api).filter(x => 1 || x.toLowerCase().indexOf('recast')!== -1));
const methods = [];
leafClasses.forEach((class_) => {
    const params = [j.identifier('node')];

    const m = j.methodDefinition('method', j.identifier(`visit_${class_}`),
                                 j.functionExpression(null, [j.identifier('node')], j.blockStatement([])));
    const eslintHint = j.commentBlock(' eslint-disable-next-line camelcase,no-unused-vars,class-methods-use-this ');

    m.comments = [eslintHint];
    const depart = j.methodDefinition('method', j.identifier(`depart_${class_}`),
                                      j.functionExpression(null, [j.identifier('node')], j.blockStatement([])));
    depart.comments = [eslintHint];
    // console.log(m);
    transBody.body.push(m, depart);
    // methods.push(m.toString());
});
fs.createWriteStream('code.js').write(recast.print(f).code);
