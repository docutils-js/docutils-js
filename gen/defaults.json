{"DocutilsCoreOptionParser":{"exitStatusLevel":5,"inputEncodingErrorHandler":"strict","outputEncoding":"utf-8","outputEncodingErrorHandler":"strict","errorEncoding":"UTF-8","errorEncodingErrorHandler":"backslashreplace","languageCode":"en","recordDependencies":null,"idPrefix":"","autoIdPrefix":"id"},"DocutilsFrontendOptionParser":{"exitStatusLevel":5,"inputEncodingErrorHandler":"strict","outputEncoding":"utf-8","outputEncodingErrorHandler":"strict","errorEncoding":"UTF-8","errorEncodingErrorHandler":"backslashreplace","languageCode":"en","recordDependencies":null,"idPrefix":"","autoIdPrefix":"id"},"DocutilsParsersRstParser":{"pepBaseUrl":"http://www.python.org/dev/peps/","pepFileUrlTemplate":"pep-%04d","rfcBaseUrl":"http://tools.ietf.org/html/","tabWidth":8,"syntaxHighlight":"long","smartQuotes":false,"characterLevelInlineMarkup":false},"DocutilsWritersDocutilsXmlWriter":{"xmlDeclaration":1,"doctypeDeclaration":1},"DocutilsWritersPepHtmlWriter":{"template":"/usr/share/docutils/writers/html4css1/template.txt","stylesheetPath":["html4css1.css"],"stylesheetDirs":[".","/usr/share/docutils/writers/html4css1","/usr/share/docutils/writers/html5_polyglot"],"initialHeaderLevel":"1","fieldNameLimit":14,"optionLimit":14,"footnoteReferences":"brackets","attribution":"dash","tableStyle":"","mathOutput":"HTML math.css","xmlDeclaration":1},"DocutilsWritersLatex2EWriter":{"documentclass":"article","documentoptions":"a4paper","docutilsFootnotes":true,"footnoteReferences":"superscript","attribution":"dash","stylesheet":"","embedStylesheet":0,"stylesheetDirs":["."],"latexPreamble":"% PDF Standard Fonts\n\\usepackage{mathptmx} % Times\n\\usepackage[scaled=.90]{helvet}\n\\usepackage{courier}","template":"default.tex","usePartSection":0,"useLatexDocinfo":0,"useLatexAbstract":0,"hyperlinkColor":"blue","hyperrefOptions":"","sectionEnumeratorSeparator":"-","literalBlockEnv":"","useVerbatimWhenPossible":0,"tableStyle":["standard"],"graphicxOption":"","fontEncoding":"T1","referenceLabel":null,"useBibtex":null},"DocutilsWritersOdfOdtWriter":{"stylesheet":"/usr/share/docutils/writers/odf_odt/styles.odt","cloakEmailAddresses":false,"tableBorderThickness":null,"addSyntaxHighlighting":false,"createSections":true,"createLinks":false,"endnotesEndDoc":false,"generateOowriterToc":true,"customHeader":"","customFooter":""},"DocutilsWritersOdfOdtReader":{"doctitleXform":1,"docinfoXform":1},"DocutilsWritersHtml4Css1Writer":{"template":"/usr/share/docutils/writers/html4css1/template.txt","stylesheetPath":["html4css1.css"],"stylesheetDirs":[".","/usr/share/docutils/writers/html4css1","/usr/share/docutils/writers/html5_polyglot"],"initialHeaderLevel":"1","fieldNameLimit":14,"optionLimit":14,"footnoteReferences":"brackets","attribution":"dash","tableStyle":"","mathOutput":"HTML math.css","xmlDeclaration":1},"DocutilsWritersXetexWriter":{"documentclass":"article","documentoptions":"a4paper","docutilsFootnotes":true,"footnoteReferences":"superscript","attribution":"dash","stylesheet":"","embedStylesheet":0,"stylesheetDirs":["."],"latexPreamble":"% Linux Libertine (free, wide coverage, not only for Linux)\n\\setmainfont{Linux Libertine O}\n\\setsansfont{Linux Biolinum O}\n\\setmonofont[HyphenChar=None,Scale=MatchLowercase]{DejaVu Sans Mono}","template":"xelatex.tex","usePartSection":0,"useLatexDocinfo":0,"useLatexAbstract":0,"hyperlinkColor":"blue","hyperrefOptions":"","sectionEnumeratorSeparator":"-","literalBlockEnv":"","useVerbatimWhenPossible":0,"tableStyle":["standard"],"graphicxOption":"","referenceLabel":null,"useBibtex":null},"DocutilsWritersHtml5PolyglotWriter":{"template":"/usr/share/docutils/writers/html5_polyglot/template.txt","stylesheetPath":["minimal.css","plain.css"],"stylesheetDirs":[".","/usr/share/docutils/writers/html5_polyglot"],"initialHeaderLevel":"1","footnoteReferences":"brackets","attribution":"dash","tableStyle":"","mathOutput":"HTML math.css"},"DocutilsWritersS5HtmlWriter":{"template":"/usr/share/docutils/writers/html4css1/template.txt","stylesheetPath":["html4css1.css"],"stylesheetDirs":[".","/usr/share/docutils/writers/html4css1","/usr/share/docutils/writers/html5_polyglot"],"initialHeaderLevel":"1","fieldNameLimit":14,"optionLimit":14,"footnoteReferences":"brackets","attribution":"dash","tableStyle":"","mathOutput":"HTML math.css","xmlDeclaration":1},"DocutilsReadersPepReader":{},"DocutilsReadersStandaloneReader":{"doctitleXform":1,"docinfoXform":1}}