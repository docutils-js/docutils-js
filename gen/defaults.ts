import {Settings} from './Settings';

const defaults: Settings = {
    "docutilsCoreOptionParser": {
        "tocBacklinks": "entry",
        "footnoteBacklinks": 1,
        "sectnumXform": 1,
        "reportLevel": 2,
        "haltLevel": 4,
        "exitStatusLevel": 5,
        "traceback": null,
        "inputEncodingErrorHandler": "strict",
        "outputEncoding": "utf-8",
        "outputEncodingErrorHandler": "strict",
        "errorEncoding": "UTF-8",
        "errorEncodingErrorHandler": "backslashreplace",
        "languageCode": "en",
        "recordDependencies": null,
        "idPrefix": "",
        "autoIdPrefix": "id"
    },
    "docutilsFrontendOptionParser": {
        "tocBacklinks": "entry",
        "footnoteBacklinks": 1,
        "sectnumXform": 1,
        "reportLevel": 2,
        "haltLevel": 4,
        "exitStatusLevel": 5,
        "traceback": null,
        "inputEncodingErrorHandler": "strict",
        "outputEncoding": "utf-8",
        "outputEncodingErrorHandler": "strict",
        "errorEncoding": "UTF-8",
        "errorEncodingErrorHandler": "backslashreplace",
        "languageCode": "en",
        "recordDependencies": null,
        "idPrefix": "",
        "autoIdPrefix": "id"
    },
    "docutilsParsersRstParser": {
        "pepBaseUrl": "http://www.python.org/dev/peps/",
        "pepFileUrlTemplate": "pep-%04d",
        "rfcBaseUrl": "http://tools.ietf.org/html/",
        "tabWidth": 8,
        "fileInsertionEnabled": 1,
        "rawEnabled": 1,
        "syntaxHighlight": "long",
        "smartQuotes": false,
        "characterLevelInlineMarkup": false
    },
    "docutilsWritersDocutilsXmlWriter": {"xmlDeclaration": true, "doctypeDeclaration": true},
    "docutilsWritersPepHtmlWriter": {
        "template": "/usr/share/docutils/writers/html4css1/template.txt",
        "stylesheetPath": ["html4css1.css"],
        "embedStylesheet": 1,
        "stylesheetDirs": [".", "/usr/share/docutils/writers/html4css1", "/usr/share/docutils/writers/html5_polyglot"],
        "initialHeaderLevel": "1",
        "fieldNameLimit": 14,
        "optionLimit": 14,
        "footnoteReferences": "brackets",
        "attribution": "dash",
        "compactLists": 1,
        "compactFieldLists": 1,
        "tableStyle": "",
        "mathOutput": "HTML math.css",
        "xmlDeclaration": true
    },
    "docutilsWritersLatex2EWriter": {
        "documentclass": "article",
        "documentoptions": "a4paper",
        "docutilsFootnotes": true,
        "footnoteReferences": "superscript",
        "useLatexCitations": false,
        "attribution": "dash",
        "stylesheet": "",
        "embedStylesheet": false,
        "stylesheetDirs": ["."],
        "latexPreamble": "% PDF Standard Fonts\n\\usepackage{mathptmx} % Times\n\\usepackage[scaled=.90]{helvet}\n\\usepackage{courier}",
        "template": "default.tex",
        "useLatexToc": true,
        "usePartSection": false,
        "useLatexDocinfo": false,
        "useLatexAbstract": false,
        "hyperlinkColor": "blue",
        "hyperrefOptions": "",
        "compoundEnumerators": null,
        "sectionPrefixForEnumerators": null,
        "sectionEnumeratorSeparator": "-",
        "literalBlockEnv": "",
        "useVerbatimWhenPossible": false,
        "tableStyle": ["standard"],
        "graphicxOption": "",
        "fontEncoding": "T1",
        "referenceLabel": null,
        "useBibtex": null
    },
    "docutilsWritersOdfOdtWriter": {
        "stylesheet": "/usr/share/docutils/writers/odf_odt/styles.odt",
        "cloakEmailAddresses": false,
        "tableBorderThickness": null,
        "addSyntaxHighlighting": false,
        "createSections": true,
        "createLinks": false,
        "endnotesEndDoc": false,
        "generateOowriterToc": true,
        "customHeader": "",
        "customFooter": ""
    },
    "docutilsWritersOdfOdtReader": {"doctitleXform": true, "docinfoXform": true, "sectsubtitleXform": false},
    "docutilsWritersHtml4Css1Writer": {
        "template": "/usr/share/docutils/writers/html4css1/template.txt",
        "stylesheetPath": ["html4css1.css"],
        "embedStylesheet": 1,
        "stylesheetDirs": [".", "/usr/share/docutils/writers/html4css1", "/usr/share/docutils/writers/html5_polyglot"],
        "initialHeaderLevel": "1",
        "fieldNameLimit": 14,
        "optionLimit": 14,
        "footnoteReferences": "brackets",
        "attribution": "dash",
        "compactLists": 1,
        "compactFieldLists": 1,
        "tableStyle": "",
        "mathOutput": "HTML math.css",
        "xmlDeclaration": true
    },
    "docutilsWritersXetexWriter": {
        "documentclass": "article",
        "documentoptions": "a4paper",
        "docutilsFootnotes": true,
        "footnoteReferences": "superscript",
        "useLatexCitations": true,
        "attribution": "dash",
        "stylesheet": "",
        "embedStylesheet": false,
        "stylesheetDirs": ["."],
        "latexPreamble": "% Linux Libertine (free, wide coverage, not only for Linux)\n\\setmainfont{Linux Libertine O}\n\\setsansfont{Linux Biolinum O}\n\\setmonofont[HyphenChar=None,Scale=MatchLowercase]{DejaVu Sans Mono}",
        "template": "xelatex.tex",
        "useLatexToc": true,
        "usePartSection": false,
        "useLatexDocinfo": false,
        "useLatexAbstract": false,
        "hyperlinkColor": "blue",
        "hyperrefOptions": "",
        "compoundEnumerators": null,
        "sectionPrefixForEnumerators": null,
        "sectionEnumeratorSeparator": "-",
        "literalBlockEnv": "",
        "useVerbatimWhenPossible": false,
        "tableStyle": ["standard"],
        "graphicxOption": "",
        "referenceLabel": null,
        "useBibtex": null
    },
    "docutilsWritersHtml5PolyglotWriter": {
        "template": "/usr/share/docutils/writers/html5_polyglot/template.txt",
        "stylesheetPath": ["minimal.css", "plain.css"],
        "embedStylesheet": 1,
        "stylesheetDirs": [".", "/usr/share/docutils/writers/html5_polyglot"],
        "initialHeaderLevel": "1",
        "footnoteReferences": "brackets",
        "attribution": "dash",
        "compactLists": true,
        "compactFieldLists": true,
        "tableStyle": "",
        "mathOutput": "HTML math.css",
        "xmlDeclaration": false
    },
    "docutilsWritersS5HtmlWriter": {
        "template": "/usr/share/docutils/writers/html4css1/template.txt",
        "stylesheetPath": ["html4css1.css"],
        "embedStylesheet": 1,
        "stylesheetDirs": [".", "/usr/share/docutils/writers/html4css1", "/usr/share/docutils/writers/html5_polyglot"],
        "initialHeaderLevel": "1",
        "fieldNameLimit": 14,
        "optionLimit": 14,
        "footnoteReferences": "brackets",
        "attribution": "dash",
        "compactLists": 1,
        "compactFieldLists": 1,
        "tableStyle": "",
        "mathOutput": "HTML math.css",
        "xmlDeclaration": true
    },
    "docutilsReadersPepReader": {},
    "docutilsReadersStandaloneReader": {"doctitleXform": true, "docinfoXform": true, "sectsubtitleXform": false}
};
;
export default defaults;
