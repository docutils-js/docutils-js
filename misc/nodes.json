{
   "footnote_reference" : {
      "leaf" : true,
      "name" : "footnote_reference",
      "classTypes" : [
         "General",
         "Inline",
         "Referential"
      ]
   },
   "address" : {
      "name" : "address",
      "classTypes" : [
         "Bibliographic"
      ],
      "leaf" : true
   },
   "literal_block" : {
      "leaf" : true,
      "name" : "literal_block",
      "classTypes" : [
         "General"
      ]
   },
   "block_quote" : {
      "leaf" : true,
      "name" : "block_quote",
      "classTypes" : [
         "General"
      ]
   },
   "field_list" : {
      "leaf" : true,
      "name" : "field_list",
      "classTypes" : [
         "Sequential"
      ]
   },
   "reference" : {
      "leaf" : true,
      "name" : "reference",
      "classTypes" : [
         "General",
         "Inline",
         "Referential"
      ]
   },
   "compound" : {
      "leaf" : true,
      "name" : "compound",
      "classTypes" : [
         "General"
      ]
   },
   "citation_reference" : {
      "name" : "citation_reference",
      "classTypes" : [
         "General",
         "Inline",
         "Referential"
      ],
      "leaf" : true
   },
   "classifier" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "classifier"
   },
   "substitution_reference" : {
      "name" : "substitution_reference",
      "classTypes" : [
         "Inline"
      ],
      "leaf" : true
   },
   "option_list_item" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "option_list_item",
      "leaf" : true
   },
   "note" : {
      "leaf" : true,
      "classTypes" : [
         "Admonition"
      ],
      "name" : "note"
   },
   "important" : {
      "leaf" : true,
      "classTypes" : [
         "Admonition"
      ],
      "name" : "important"
   },
   "comment" : {
      "leaf" : true,
      "name" : "comment",
      "classTypes" : [
         "Special",
         "Invisible",
         "Inline",
         "Targetable"
      ]
   },
   "bullet_list" : {
      "classTypes" : [
         "Sequential"
      ],
      "name" : "bullet_list",
      "leaf" : true
   },
   "table" : {
      "classTypes" : [
         "General"
      ],
      "name" : "table",
      "leaf" : true
   },
   "container" : {
      "leaf" : true,
      "name" : "container",
      "classTypes" : [
         "General"
      ]
   },
   "version" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "version",
      "leaf" : true
   },
   "hint" : {
      "leaf" : true,
      "classTypes" : [
         "Admonition"
      ],
      "name" : "hint"
   },
   "revision" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "revision",
      "leaf" : true
   },
   "citation" : {
      "classTypes" : [
         "General",
         "BackLinkable",
         "Labeled",
         "Targetable"
      ],
      "name" : "citation",
      "leaf" : true
   },
   "subtitle" : {
      "name" : "subtitle",
      "classTypes" : [
         "Titular",
         "PreBibliographic"
      ],
      "leaf" : true
   },
   "line" : {
      "leaf" : true,
      "name" : "line",
      "classTypes" : [
         "Part"
      ]
   },
   "emphasis" : {
      "name" : "emphasis",
      "classTypes" : [
         "Inline"
      ],
      "leaf" : true
   },
   "footer" : {
      "classTypes" : [
         "Decorative"
      ],
      "name" : "footer",
      "leaf" : true
   },
   "definition_list" : {
      "classTypes" : [
         "Sequential"
      ],
      "name" : "definition_list",
      "leaf" : true
   },
   "footnote" : {
      "leaf" : true,
      "name" : "footnote",
      "classTypes" : [
         "General",
         "BackLinkable",
         "Labeled",
         "Targetable"
      ]
   },
   "author" : {
      "name" : "author",
      "classTypes" : [
         "Bibliographic"
      ],
      "leaf" : true
   },
   "row" : {
      "name" : "row",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "figure" : {
      "leaf" : true,
      "classTypes" : [
         "General"
      ],
      "name" : "figure"
   },
   "docinfo" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "docinfo",
      "leaf" : true
   },
   "copyright" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "copyright",
      "leaf" : true
   },
   "title_reference" : {
      "classTypes" : [
         "Inline"
      ],
      "name" : "title_reference",
      "leaf" : true
   },
   "math_block" : {
      "leaf" : true,
      "name" : "math_block",
      "classTypes" : [
         "General"
      ]
   },
   "acronym" : {
      "leaf" : true,
      "classTypes" : [
         "Inline"
      ],
      "name" : "acronym"
   },
   "superscript" : {
      "classTypes" : [
         "Inline"
      ],
      "name" : "superscript",
      "leaf" : true
   },
   "system_message" : {
      "classTypes" : [
         "Special",
         "BackLinkable",
         "PreBibliographic"
      ],
      "name" : "system_message",
      "leaf" : true
   },
   "substitution_definition" : {
      "leaf" : true,
      "classTypes" : [
         "Special",
         "Invisible"
      ],
      "name" : "substitution_definition"
   },
   "error" : {
      "leaf" : true,
      "name" : "error",
      "classTypes" : [
         "Admonition"
      ]
   },
   "option_group" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "option_group"
   },
   "danger" : {
      "leaf" : true,
      "name" : "danger",
      "classTypes" : [
         "Admonition"
      ]
   },
   "target" : {
      "name" : "target",
      "classTypes" : [
         "Special",
         "Invisible",
         "Inline",
         "Targetable"
      ],
      "leaf" : true
   },
   "attention" : {
      "leaf" : true,
      "name" : "attention",
      "classTypes" : [
         "Admonition"
      ]
   },
   "admonition" : {
      "name" : "admonition",
      "classTypes" : [
         "Admonition"
      ],
      "leaf" : true
   },
   "topic" : {
      "name" : "topic",
      "classTypes" : [
         "Structural"
      ],
      "leaf" : true
   },
   "paragraph" : {
      "leaf" : true,
      "classTypes" : [
         "General"
      ],
      "name" : "paragraph"
   },
   "rubric" : {
      "leaf" : true,
      "name" : "rubric",
      "classTypes" : [
         "Titular"
      ]
   },
   "label" : {
      "leaf" : true,
      "name" : "label",
      "classTypes" : [
         "Part"
      ]
   },
   "Element" : {
      "subClasses" : [
         "header",
         "footer",
         "decoration",
         "TextElement",
         "document",
         "docinfo",
         "authors",
         "section",
         "topic",
         "sidebar",
         "transition",
         "compound",
         "container",
         "bullet_list",
         "enumerated_list",
         "list_item",
         "definition_list",
         "definition_list_item",
         "definition",
         "field_list",
         "field",
         "field_body",
         "option",
         "option_group",
         "option_list",
         "option_list_item",
         "description",
         "line_block",
         "block_quote",
         "attention",
         "caution",
         "danger",
         "error",
         "important",
         "note",
         "tip",
         "hint",
         "warning",
         "admonition",
         "footnote",
         "citation",
         "figure",
         "legend",
         "table",
         "tgroup",
         "colspec",
         "thead",
         "tbody",
         "row",
         "entry",
         "system_message",
         "pending",
         "image"
      ],
      "name" : "Element",
      "classTypes" : [],
      "leaf" : false
   },
   "subscript" : {
      "classTypes" : [
         "Inline"
      ],
      "name" : "subscript",
      "leaf" : true
   },
   "document" : {
      "leaf" : true,
      "classTypes" : [
         "Root",
         "Structural"
      ],
      "name" : "document"
   },
   "field_name" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "field_name",
      "leaf" : true
   },
   "sidebar" : {
      "name" : "sidebar",
      "classTypes" : [
         "Structural"
      ],
      "leaf" : true
   },
   "authors" : {
      "leaf" : true,
      "name" : "authors",
      "classTypes" : [
         "Bibliographic"
      ]
   },
   "tip" : {
      "leaf" : true,
      "classTypes" : [
         "Admonition"
      ],
      "name" : "tip"
   },
   "attribution" : {
      "name" : "attribution",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "generated" : {
      "leaf" : true,
      "name" : "generated",
      "classTypes" : [
         "Inline"
      ]
   },
   "math" : {
      "leaf" : true,
      "classTypes" : [
         "Inline"
      ],
      "name" : "math"
   },
   "option" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "option",
      "leaf" : true
   },
   "entry" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "entry",
      "leaf" : true
   },
   "legend" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "legend",
      "leaf" : true
   },
   "warning" : {
      "leaf" : true,
      "name" : "warning",
      "classTypes" : [
         "Admonition"
      ]
   },
   "description" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "description"
   },
   "date" : {
      "name" : "date",
      "classTypes" : [
         "Bibliographic"
      ],
      "leaf" : true
   },
   "list_item" : {
      "name" : "list_item",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "option_argument" : {
      "name" : "option_argument",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "tgroup" : {
      "name" : "tgroup",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "strong" : {
      "leaf" : true,
      "classTypes" : [
         "Inline"
      ],
      "name" : "strong"
   },
   "literal" : {
      "leaf" : true,
      "name" : "literal",
      "classTypes" : [
         "Inline"
      ]
   },
   "caution" : {
      "leaf" : true,
      "classTypes" : [
         "Admonition"
      ],
      "name" : "caution"
   },
   "section" : {
      "leaf" : true,
      "classTypes" : [
         "Structural"
      ],
      "name" : "section"
   },
   "image" : {
      "leaf" : true,
      "name" : "image",
      "classTypes" : [
         "General",
         "Inline"
      ]
   },
   "header" : {
      "leaf" : true,
      "name" : "header",
      "classTypes" : [
         "Decorative"
      ]
   },
   "definition_list_item" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "definition_list_item"
   },
   "decoration" : {
      "classTypes" : [
         "Decorative"
      ],
      "name" : "decoration",
      "leaf" : true
   },
   "Node" : {
      "subClasses" : [
         "Element",
         "Text"
      ],
      "name" : "Node",
      "classTypes" : [],
      "leaf" : false
   },
   "field" : {
      "name" : "field",
      "classTypes" : [
         "Part"
      ],
      "leaf" : true
   },
   "option_string" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "option_string"
   },
   "field_body" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "field_body"
   },
   "raw" : {
      "name" : "raw",
      "classTypes" : [
         "Special",
         "Inline",
         "PreBibliographic"
      ],
      "leaf" : true
   },
   "doctest_block" : {
      "leaf" : true,
      "classTypes" : [
         "General"
      ],
      "name" : "doctest_block"
   },
   "inline" : {
      "name" : "inline",
      "classTypes" : [
         "Inline"
      ],
      "leaf" : true
   },
   "transition" : {
      "leaf" : true,
      "classTypes" : [
         "Structural"
      ],
      "name" : "transition"
   },
   "thead" : {
      "leaf" : true,
      "name" : "thead",
      "classTypes" : [
         "Part"
      ]
   },
   "status" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "status",
      "leaf" : true
   },
   "pending" : {
      "leaf" : true,
      "classTypes" : [],
      "name" : "pending"
   },
   "colspec" : {
      "leaf" : true,
      "name" : "colspec",
      "classTypes" : [
         "Part"
      ]
   },
   "tbody" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "tbody"
   },
   "title" : {
      "leaf" : true,
      "name" : "title",
      "classTypes" : [
         "Titular",
         "PreBibliographic"
      ]
   },
   "problematic" : {
      "name" : "problematic",
      "classTypes" : [
         "Inline"
      ],
      "leaf" : true
   },
   "term" : {
      "leaf" : true,
      "classTypes" : [
         "Part"
      ],
      "name" : "term"
   },
   "organization" : {
      "leaf" : true,
      "name" : "organization",
      "classTypes" : [
         "Bibliographic"
      ]
   },
   "TextElement" : {
      "classTypes" : [],
      "name" : "TextElement",
      "subClasses" : [
         "FixedTextElement",
         "title",
         "subtitle",
         "rubric",
         "author",
         "organization",
         "contact",
         "version",
         "revision",
         "status",
         "date",
         "copyright",
         "paragraph",
         "term",
         "classifier",
         "field_name",
         "option_argument",
         "option_string",
         "line",
         "attribution",
         "substitution_definition",
         "target",
         "label",
         "caption",
         "emphasis",
         "strong",
         "literal",
         "reference",
         "footnote_reference",
         "citation_reference",
         "substitution_reference",
         "title_reference",
         "abbreviation",
         "acronym",
         "superscript",
         "subscript",
         "math",
         "inline",
         "problematic",
         "generated"
      ],
      "leaf" : false
   },
   "caption" : {
      "classTypes" : [
         "Part"
      ],
      "name" : "caption",
      "leaf" : true
   },
   "option_list" : {
      "classTypes" : [
         "Sequential"
      ],
      "name" : "option_list",
      "leaf" : true
   },
   "abbreviation" : {
      "name" : "abbreviation",
      "classTypes" : [
         "Inline"
      ],
      "leaf" : true
   },
   "Text" : {
      "classTypes" : [],
      "name" : "Text",
      "leaf" : true
   },
   "contact" : {
      "classTypes" : [
         "Bibliographic"
      ],
      "name" : "contact",
      "leaf" : true
   },
   "FixedTextElement" : {
      "leaf" : false,
      "subClasses" : [
         "address",
         "literal_block",
         "doctest_block",
         "math_block",
         "comment",
         "raw"
      ],
      "name" : "FixedTextElement",
      "classTypes" : []
   },
   "line_block" : {
      "leaf" : true,
      "classTypes" : [
         "General"
      ],
      "name" : "line_block"
   },
   "definition" : {
      "leaf" : true,
      "name" : "definition",
      "classTypes" : [
         "Part"
      ]
   },
   "enumerated_list" : {
      "classTypes" : [
         "Sequential"
      ],
      "name" : "enumerated_list",
      "leaf" : true
   }
}
