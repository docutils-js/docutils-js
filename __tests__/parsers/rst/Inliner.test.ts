// @ts-ignore
import formatXml from 'xml-formatter';
import Inliner from '../../../src/parsers/rst/Inliner';
import newReporter from '../../../src/newReporter';
import newDocument from '../../../src/newDocument';
import { nodeToXml } from '../../../src/nodes';
import defaults from "../../../gen/defaults";

const currentLogLines: string[] = [];

afterEach(() => {
    if (currentLogLines.length) {
        /* eslint-disable-next-line no-unused-vars */
        currentLogLines.forEach((line) => {});
        currentLogLines.length = 0;
    }
});

test('inliner 1', () => {
    const inliner = new Inliner();
    inliner.initCustomizations({...defaults});
    const document = newDocument({ sourcePath: ''}, { ...defaults});
    const reporter = newReporter({ sourcePath: ''}, { ...defaults});
    let language;
    const memo = {
        document,
        reporter,
        language,
    };

    const result = inliner.parse('`test`:foo:', { lineno: 1, memo, parent: document });
    const [nodes, messages] = result;
    expect(nodes.map(n => formatXml(nodeToXml(n))).join('')).toMatchSnapshot();
    expect(messages.map(n => formatXml(nodeToXml(n))).join('')).toMatchSnapshot();
});

test.each([['Interpreted text', '`test`:foo:'],
           ['Emphasis', 'I like *TV*'],
           ['Bold', 'Eat **lots** of *food*.'],
           ['Literal', '``literal``'],
           ['Not sure', '_`hello`'],
           ['Interpreted text, no specified role', '`test`'],
          ])('%s', (testName, a) => {
    const inliner = new Inliner();
    inliner.initCustomizations({...defaults});
              const document = newDocument({sourcePath:''}, { ...defaults });
              const reporter = newReporter({ sourcePath:''}, { ...defaults});
    let language;
    const memo = {
 document,
                   reporter,
                   language,
                 };

    const result = inliner.parse(a, { lineno: 1, memo, parent: document });
              const [nodes, messages] = result;
              expect(nodes.map(n => formatXml(nodeToXml(n))).join('')).toMatchSnapshot();
              expect(messages.map(n => formatXml(nodeToXml(n))).join('')).toMatchSnapshot();
});
