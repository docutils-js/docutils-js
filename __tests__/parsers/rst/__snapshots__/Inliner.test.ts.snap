// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bold 1`] = `""`;

exports[`Bold 2`] = `""`;

exports[`Emphasis 1`] = `""`;

exports[`Emphasis 2`] = `""`;

exports[`Interpreted text 1`] = `""`;

exports[`Interpreted text 2`] = `""`;

exports[`Interpreted text, no specified role 1`] = `""`;

exports[`Interpreted text, no specified role 2`] = `""`;

exports[`Literal 1`] = `""`;

exports[`Literal 2`] = `""`;

exports[`Not sure 1`] = `""`;

exports[`Not sure 2`] = `""`;

exports[`inliner 1 1`] = `""`;

exports[`inliner 1 2`] = `""`;
